src/SDK/AttributesBuilder.php

```diff
@@ -0,0 +1,134 @@
+use function mb_substr;
+
+/**
+ * @internal
+ */
+final class AttributesBuilder implements AttributesBuilderInterface
+{
+    private array $attributes;
+    private ?int $attributeCountLimit;
+    private ?int $attributeValueLengthLimit;
+    private int $droppedAttributesCount;
+
+    private function __construct(array $attributes, ?int $attributeCountLimit, ?int $attributeValueLengthLimit, int $droppedAttributesCount)
+    {
+        $this->attributes = $attributes;
+        $this->attributeCountLimit = $attributeCountLimit;
+        $this->attributeValueLengthLimit = $attributeValueLengthLimit;
+        $this->droppedAttributesCount = $droppedAttributesCount;
+    }
+
+    public static function from(iterable $attributes, ?int $attributeCountLimit = null, ?int $attributeValueLengthLimit = null): AttributesBuilder
+    {
+        if (!$attributes) {
+            return new self([], $attributeCountLimit, $attributeValueLengthLimit, 0);
+        }
+
+        $droppedAttributesCount = $attributes instanceof AttributesInterface
+            ? $attributes->getDroppedAttributesCount()
+            : 0;
+
+        if ($attributes instanceof Attributes) {
+            $attributes = $attributes->toArray();
+        } elseif (is_array($attributes)) {
+            foreach ($attributes as $key => $value) {
+                if ($value === null) {
+                    unset($attributes[$key]);
+                }
+            }
+        }
+
+        if (is_array($attributes)) {
+            if ($attributeCountLimit !== null && count($attributes) > $attributeCountLimit) {
+                $droppedAttributesCount += count($attributes) - $attributeCountLimit;
+                array_splice($attributes, $attributeCountLimit);
+            }
+            if ($attributeValueLengthLimit !== null) {
+                foreach ($attributes as $key => $value) {
+                    $limited = self::applyLengthLimit($value, $attributeValueLengthLimit);
+                    if ($limited === $value) {
+                        continue;
+                    }
+
+                    $attributes[$key] = $limited;
+                }
+            }
+
+            return new self($attributes, $attributeCountLimit, $attributeValueLengthLimit, $droppedAttributesCount);
+        }
+
+        $sdkAttributes = new self([], $attributeCountLimit, $attributeValueLengthLimit, $droppedAttributesCount);
+        foreach ($attributes as $key => $value) {
+            $sdkAttributes[$key] = $value;
+        }
+
+        return $sdkAttributes;
+    }
+
+    public function build(): AttributesInterface
+    {
+        return new Attributes($this->attributes, $this->droppedAttributesCount);
+    }
+
+    public function offsetExists($offset): bool
+    {
+        return isset($this->attributes[$offset]);
+    }
+
+    public function offsetGet($offset)
+    {
+        return $this->attributes[$offset] ?? null;
+    }
+
+    public function offsetSet($offset, $value): void
```

Comment:
An `$offset` which is an Attribute key "must be a non-null and non-empty string" ([spec](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/common/common.md#attributes)). It's better to add type hint `string`.

Reply:
Also, needs to be compatible with ArrayAccess signature for 8.1+, see https://github.com/open-telemetry/opentelemetry-php/runs/4729027088?check_suite_focus=true#step:13:45